exporters/otlp/otlpgrpc/mock_collector_test.go

```diff
@@ -189,19 +205,62 @@ func runMockCollectorAtEndpoint(t *testing.T, endpoint string) *mockCollector {
 	mc := makeMockCollector(t)
 	collectortracepb.RegisterTraceServiceServer(srv, mc.traceSvc)
 	collectormetricpb.RegisterMetricsServiceServer(srv, mc.metricSvc)
+	mc.ln = newListener(ln)
 	go func() {
-		_ = srv.Serve(ln)
+		_ = srv.Serve((net.Listener)(mc.ln))
 	}()
 
-	deferFunc := func() error {
-		srv.Stop()
-		return ln.Close()
+	mc.endpoint = ln.Addr().String()
+	// srv.Stop calls Close on mc.ln.
+	mc.stopFunc = srv.Stop
+
+	return mc
+}
+
+type listener struct {
+	wrapped net.Listener
+
+	C chan struct{}
+
+	closed    chan struct{}
+	closeOnce sync.Once
+}
+
+func newListener(wrapped net.Listener) *listener {
+	return &listener{
+		wrapped: wrapped,
+		C:       make(chan struct{}, 1),
+		closed:  make(chan struct{}),
 	}
+}
 
-	_, collectorPortStr, _ := net.SplitHostPort(ln.Addr().String())
+func (l *listener) Accept() (net.Conn, error) {
+	conn, err := l.wrapped.Accept()
```

Comment:
This should return an error when wrapped is closed.  Is it possible to use that as the signal that we have called closed?